### PR TITLE
Associated types for Corpus, State

### PR DESCRIPTION
**Pull request by domenukk**
A less ambitious but actually working PR superseding #749

### CODE REVIEW

**Comment by domenukk** [comment]
@andreafioraldi @tokatoka @s1341 continue or abort?

**Review by s1341 — APPROVED** [review]
lgtm! I'd keep going... I think the goal is a very worthy one!

File: bindings/pylibafl/test.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def is_interesting(self, state, mgr, input, observers, exit_kind):
         return ob.n % 10000 == 0
 
 class FooExecutor(BaseExecutor):
-    def __init__(self, harness, observers: ObserversTuple):
+    def __init__(self, harness, observers: ObserversTuple<Input = I, State =S>):
```
Comment (andreafioraldi) [inline bindings/pylibafl/test.py:20]: hemn

Comment (domenukk) [reply bindings/pylibafl/test.py:20]: That's a bug 😅

Comment (domenukk) [inline bindings/pylibafl/test.py:20]: Bug

**Comment by domenukk** [comment]
@VTCAKAVSMoACE to not run into the prelude problems, prelude is a feature flag that you can disable for $toolofyourchoice

**Comment by VTCAKAVSMoACE** [comment]
Good tip. Fixed.

**Comment by domenukk** [comment]
Getting the python bindings to work again may be a bit of a pain with this
Also once it compiles, we may be able to remove a lot of complexity again, at least in my head

**Comment by VTCAKAVSMoACE** [comment]
Yeah I'm having a "specialist" (my friend @nhwn who is a generics/at wizard) take a peek.

**Comment by VTCAKAVSMoACE** [comment]
@domenukk Okay, kind of a dumb fix, but I just de-associated State from Feedback. Inference works again; we definitely hit a known Rust limitation.

File: scripts/test_all_fuzzers.sh [hunk 2]
```diff
@@ -67,7 +67,7 @@ do
 	    echo "[+] Done testing $fuzzer"
     else
         echo "[*] Building $fuzzer"
-        cargo build || exit 1
+        cargo build --target-dir ./target || exit 1
```
Comment (Mrmaxmeier) [inline scripts/test_all_fuzzers.sh:70]: Good call, but I don't think this is doing anything. (`--target-dir=target` is the default)
I've opened https://github.com/AFLplusplus/LibAFL/pull/845 to test shared target dirs. :)

**Comment by domenukk** [comment]
I don't think loose bounds are an issue tbh? It makes the whole thing more flexible, and I don't see huge drawbacks, right?

**Comment by VTCAKAVSMoACE** [comment]
> I don't think loose bounds are an issue tbh? It makes the whole thing more flexible, and I don't see huge drawbacks, right?

It is, primarily because people will model their code after this. Any custom implementations are going to be based on existing ones; we want to demonstrate the easiest method. It won't be too much extra work.

**Comment by domenukk** [comment]
Why would loose bounds lead to "less good" code?

**Comment by VTCAKAVSMoACE** [comment]
It's possible to constrain further (e.g., enforce consistency between types in Executors vs EventManagers vs Fuzzers) but I can't seem to actually do this without ICEing. Fun!

File: libafl/src/corpus/minimizer.rs [hunk 3]
```diff
@@ -12,99 +12,97 @@ use num_traits::ToPrimitive;
...
+impl<E, O, T, TS> MapCorpusMinimizer<E, O, T, TS>
 where
-    E: Copy + Hash + Eq,
-    I: Input,
-    for<'a> O: MapObserver<Entry = E> + AsIter<'a, Item = E>,
-    S: HasMetadata + HasCorpus<I>,
-    TS: TestcaseScore<I, S>,
+    E: UsesState,
+    E::State: HasCorpus + HasMetadata,
+    TS: TestcaseScore<E::State>,
```
Comment (VTCAKAVSMoACE) [inline libafl/src/corpus/minimizer.rs:71]: This type was completely incorrect. Apparently, when I wrote it, I got confused between E (for Entry, in MapObserver) and E (for Executor). As a result, it was deeply underconstrained and completely incorrect. The AT changes prevented the type from compiling and provided a very useful error message.

**Comment by domenukk** [comment]
We should probably directly change the version to 0.9 so ppl don't get overly confused, right?

**Comment by domenukk** [comment]
and maybe add a tiny bit of migration help, like linking to one generics change here and explaining Uses

**Comment by domenukk** [comment]
Sorry, didn't mean to close

**Comment by VTCAKAVSMoACE** [comment]
lmao all good, it happens -- close with comment needs to be on the other side of the comment box

File: libafl/src/events/mod.rs [hunk 4]
```diff
@@ -434,18 +433,15 @@ pub trait EventRestarter<S> {
...
 
 /// [`EventProcessor`] process all the incoming messages
-pub trait EventProcessor<E, I, S, Z> {
+pub trait EventProcessor<E, Z>: UsesState {
     /// Lookup for incoming events and process them.
     /// Return the number of processes events or an error
-    fn process(&mut self, fuzzer: &mut Z, state: &mut S, executor: &mut E) -> Result<usize, Error>;
-
-    /// Deserialize all observers for this type and manager
-    fn deserialize_observers<OT>(&mut self, observers_buf: &[u8]) -> Result<OT, Error>
```
Comment (VTCAKAVSMoACE) [inline libafl/src/events/mod.rs:443]: This function was literally never used, anywhere.

**Comment by domenukk** [comment]
Compiler/ Clippy is still unhappy 🥲

**Comment by VTCAKAVSMoACE** [comment]
oh thank god, the rust cache

**Comment by VTCAKAVSMoACE** [comment]
Assuming that the fuzzers build, this is good to go! :rocket:

**Comment by domenukk** [comment]
_HUGE_ thank you! This is amazing

**Pull request merged**